v-next/core/src/internal/plugins/plugin-validation.ts

```diff
@@ -1,6 +1,101 @@
+import { createRequire } from "node:module";
+import path from "node:path";
+import process from "node:process";
+
+import { HardhatError } from "@nomicfoundation/hardhat-errors";
+import { PackageJson } from "@nomicfoundation/hardhat-utils/package";
+import semver from "semver";
+
 import { HardhatPlugin } from "../../types/plugins.js";
 
-export async function validatePluginNpmDependencies(_plugin: HardhatPlugin) {
-  // TODO: If it has an npm package, validate their peer dependencies
-  // If any is missing, throw
+/**
+ * Validate that a plugin is installed and that its peer dependencies are installed and satisfy the version constraints.
+ *
+ * @param plugin - the plugin to be validated
+ * @param basePathForNpmResolution - the directory path to use for node module resolution, defaulting to `process.cwd()`
+ */
+export async function validatePluginNpmDependencies(
+  plugin: HardhatPlugin,
+  // TODO: When all loads where based on a hardhat config file
+  // on disk we used the config file's directory as the base path
+  // We need to decide if the current working directory is the right
+  // option for module resolution now
+  basePathForNpmResolution?: string,
+): Promise<void> {
+  if (plugin.npmPackage === undefined) {
+    return;
+  }
+
+  const pluginPackageJson = readPackageJsonViaNodeRequire(
+    plugin.npmPackage,
+    basePathForNpmResolution,
+  );
+
+  if (pluginPackageJson === undefined) {
+    throw new HardhatError(HardhatError.ERRORS.PLUGINS.PLUGIN_NOT_INSTALLED, {
+      pluginId: plugin.id,
+    });
+  }
+
+  if (pluginPackageJson.peerDependencies === undefined) {
+    return;
+  }
+
+  for (const [dependencyName, versionSpec] of Object.entries(
+    pluginPackageJson.peerDependencies,
+  )) {
+    const dependencyPackageJson = readPackageJsonViaNodeRequire(
+      dependencyName,
+      basePathForNpmResolution,
```

Comment:
I think the base path here should be the directory of the plugin's package.json

Reply:
The reason for this is that it's the plugin's dependency. Normally it would also resolve from CWD, but to be correct, it should be resolvable from the plugin itself.

Reply:
I have made this change and applied a test case with `./node_modules/example/node_modules/peer1` that fails if the original version is kept in place.